### PR TITLE
OCPBUGS-10387: label for infra nodes for metric cluster:capacity_cpu_cores:sum

### PR DESCRIPTION
**Pull request by raptorsun**
As stated in [OCPBUGS-10387](https://issues.redhat.com/browse/OCPBUGS-10387), certain clusters do not label their infra nodes in the metric `cluster:capacity_cpu_cores:sum`, due to lack of label `label_node_role_kubernetes_io="infra"` in the metric `kube_node_labels`.
This PR add logic to automatically add a label  `label_node_role_kubernetes_io="infra"` to infra nodes when generating the metric `metric cluster:capacity_cpu_cores:sum`.

This PR is based on the PR https://github.com/openshift/cluster-monitoring-operator/pull/1926 that adds unit tests for the current recording rule of  `cluster:capacity_cpu_cores:sum`.

* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

**Comment by juzhao** [comment]
given we have 3 masters, 3 workers, each node is with 4 cpus, no infra node
```
$ oc get node
ocNAME                                         STATUS   ROLES                  AGE   VERSION
ip-10-0-139-233.us-east-2.compute.internal   Ready    control-plane,master   71m   v1.26.2+dc93b13
ip-10-0-146-211.us-east-2.compute.internal   Ready    worker                 64m   v1.26.2+dc93b13
ip-10-0-179-162.us-east-2.compute.internal   Ready    control-plane,master   71m   v1.26.2+dc93b13
ip-10-0-182-82.us-east-2.compute.internal    Ready    worker                 62m   v1.26.2+dc93b13
ip-10-0-195-163.us-east-2.compute.internal   Ready    worker                 62m   v1.26.2+dc93b13
ip-10-0-222-248.us-east-2.compute.internal   Ready    control-plane,master   71m   v1.26.2+dc93b13
```
search with the new expr
```
label_beta_kubernetes_io_instance_type	label_kubernetes_io_arch	label_node_openshift_io_os_id	label_node_role_kubernetes_io	Value
m5.xlarge				amd64				rhcos				master				12
m5.xlarge				amd64				rhcos								12
```
label one node ip-10-0-195-163.us-east-2.compute.internal with node-role.kubernetes.io/infra: "", the node is also worker and infra node
```
$ oc get node
NAME                                         STATUS   ROLES                  AGE    VERSION
ip-10-0-139-233.us-east-2.compute.internal   Ready    control-plane,master   145m   v1.26.2+dc93b13
ip-10-0-146-211.us-east-2.compute.internal   Ready    worker                 139m   v1.26.2+dc93b13
ip-10-0-179-162.us-east-2.compute.internal   Ready    control-plane,master   145m   v1.26.2+dc93b13
ip-10-0-182-82.us-east-2.compute.internal    Ready    worker                 136m   v1.26.2+dc93b13
ip-10-0-195-163.us-east-2.compute.internal   Ready    infra,worker           136m   v1.26.2+dc93b13
ip-10-0-222-248.us-east-2.compute.internal   Ready    control-plane,master   145m   v1.26.2+dc93b13
```
could find ip-10-0-195-163.us-east-2.compute.internal in cluster:infra_nodes metrics
```
$ token=`oc create token prometheus-k8s -n openshift-monitoring`
$ oc -n openshift-monitoring exec -c prometheus prometheus-k8s-0 -- curl -k -H "Authorization: Bearer $token" 'https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query?' --data-urlencode 'query=cluster:infra_nodes' | jq
{
  "status": "success",
  "data": {
    "resultType": "vector",
    "result": [
      {
        "metric": {
          "__name__": "cluster:infra_nodes",
          "container": "kube-rbac-proxy-main",
          "label_beta_kubernetes_io_arch": "amd64",
          "label_beta_kubernetes_io_instance_type": "m5.xlarge",
          "label_beta_kubernetes_io_os": "linux",
          "label_failure_domain_beta_kubernetes_io_region": "us-east-2",
          "label_failure_domain_beta_kubernetes_io_zone": "us-east-2c",
          "label_kubernetes_io_arch": "amd64",
          "label_kubernetes_io_hostname": "ip-10-0-195-163.us-east-2.compute.internal",
          "label_kubernetes_io_os": "linux",
          "label_node_kubernetes_io_instance_type": "m5.xlarge",
          "label_node_openshift_io_os_id": "rhcos",
          "label_node_role_kubernetes_io_infra": "true",
          "label_topology_ebs_csi_aws_com_zone": "us-east-2c",
          "label_topology_kubernetes_io_region": "us-east-2",
          "label_topology_kubernetes_io_zone": "us-east-2c",
          "namespace": "openshift-monitoring",
          "node": "ip-10-0-195-163.us-east-2.compute.internal",
          "prometheus": "openshift-monitoring/k8s"
        },
        "value": [
          1680144507.25,
          "1"
        ]
      }
    ]
  }
}

```
search with the new expr again, will get
```
label_beta_kubernetes_io_instance_type	label_kubernetes_io_arch	label_node_openshift_io_os_id	label_node_role_kubernetes_io	Value
m5.xlarge				amd64				rhcos				master				12
m5.xlarge				amd64				rhcos				infra				4
m5.xlarge				amd64				rhcos								8
```
in this case, cpu for node ip-10-0-195-163.us-east-2.compute.internal is only calculated as infra node, but not calculated as worker node

**Comment by raptorsun** [comment]
> given we have 3 masters, 3 workers, each node is with 4 cpus, no infra node
> 
> ```
> $ oc get node
> ocNAME                                         STATUS   ROLES                  AGE   VERSION
> ip-10-0-139-233.us-east-2.compute.internal   Ready    control-plane,master   71m   v1.26.2+dc93b13
> ip-10-0-146-211.us-east-2.compute.internal   Ready    worker                 64m   v1.26.2+dc93b13
> ip-10-0-179-162.us-east-2.compute.internal   Ready    control-plane,master   71m   v1.26.2+dc93b13
> ip-10-0-182-82.us-east-2.compute.internal    Ready    worker                 62m   v1.26.2+dc93b13
> ip-10-0-195-163.us-east-2.compute.internal   Ready    worker                 62m   v1.26.2+dc93b13
> ip-10-0-222-248.us-east-2.compute.internal   Ready    control-plane,master   71m   v1.26.2+dc93b13
> ```
> 
> search with the new expr, result is expected
> 
> ```
> label_beta_kubernetes_io_instance_type	label_kubernetes_io_arch	label_node_openshift_io_os_id	label_node_role_kubernetes_io	Value
> m5.xlarge				amd64				rhcos				master				12
> m5.xlarge				amd64				rhcos								12
> ```
> 
> label one node ip-10-0-195-163.us-east-2.compute.internal with node-role.kubernetes.io/infra: "", the node is also worker and infra node
> 
> ```
> $ oc get node
> NAME                                         STATUS   ROLES                  AGE    VERSION
> ip-10-0-139-233.us-east-2.compute.internal   Ready    control-plane,master   145m   v1.26.2+dc93b13
> ip-10-0-146-211.us-east-2.compute.internal   Ready    worker                 139m   v1.26.2+dc93b13
> ip-10-0-179-162.us-east-2.compute.internal   Ready    control-plane,master   145m   v1.26.2+dc93b13
> ip-10-0-182-82.us-east-2.compute.internal    Ready    worker                 136m   v1.26.2+dc93b13
> ip-10-0-195-163.us-east-2.compute.internal   Ready    infra,worker           136m   v1.26.2+dc93b13
> ip-10-0-222-248.us-east-2.compute.internal   Ready    control-plane,master   145m   v1.26.2+dc93b13
> ```
> 
> could find ip-10-0-195-163.us-east-2.compute.internal in cluster:infra_nodes metrics
> 
> ```
> $ token=`oc create token prometheus-k8s -n openshift-monitoring`
> $ oc -n openshift-monitoring exec -c prometheus prometheus-k8s-0 -- curl -k -H "Authorization: Bearer $token" 'https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query?' --data-urlencode 'query=cluster:infra_nodes' | jq
> {
>   "status": "success",
>   "data": {
>     "resultType": "vector",
>     "result": [
>       {
>         "metric": {
>           "__name__": "cluster:infra_nodes",
>           "container": "kube-rbac-proxy-main",
>           "label_beta_kubernetes_io_arch": "amd64",
>           "label_beta_kubernetes_io_instance_type": "m5.xlarge",
>           "label_beta_kubernetes_io_os": "linux",
>           "label_failure_domain_beta_kubernetes_io_region": "us-east-2",
>           "label_failure_domain_beta_kubernetes_io_zone": "us-east-2c",
>           "label_kubernetes_io_arch": "amd64",
>           "label_kubernetes_io_hostname": "ip-10-0-195-163.us-east-2.compute.internal",
>           "label_kubernetes_io_os": "linux",
>           "label_node_kubernetes_io_instance_type": "m5.xlarge",
>           "label_node_openshift_io_os_id": "rhcos",
>           "label_node_role_kubernetes_io_infra": "true",
>           "label_topology_ebs_csi_aws_com_zone": "us-east-2c",
>           "label_topology_kubernetes_io_region": "us-east-2",
>           "label_topology_kubernetes_io_zone": "us-east-2c",
>           "namespace": "openshift-monitoring",
>           "node": "ip-10-0-195-163.us-east-2.compute.internal",
>           "prometheus": "openshift-monitoring/k8s"
>         },
>         "value": [
>           1680144507.25,
>           "1"
>         ]
>       }
>     ]
>   }
> }
> ```
> 
> search with the new expr again, will get
> 
> ```
> label_beta_kubernetes_io_instance_type	label_kubernetes_io_arch	label_node_openshift_io_os_id	label_node_role_kubernetes_io	Value
> m5.xlarge				amd64				rhcos				master				12
> m5.xlarge				amd64				rhcos				infra				4
> m5.xlarge				amd64				rhcos								8
> ```
> 
> in this case, cpu for node ip-10-0-195-163.us-east-2.compute.internal is only calculated as infra node, but not calculated as worker node

Thank you for finding that problem!
The node labeled as both a worker and an infra should fall into the worker's category. Let me fix that and submit a new version later.

**Comment by juzhao** [comment]
it seems the new expr is wrong, one node is both infra/worker node
```
$ oc get node -l node-role.kubernetes.io/infra
NAME                                         STATUS   ROLES          AGE    VERSION
ip-10-0-217-149.us-east-2.compute.internal   Ready    infra,worker   129m   v1.26.2+dc93b13
```
search with the new expr, no infra result
```
label_beta_kubernetes_io_instance_type	label_kubernetes_io_arch	label_node_openshift_io_os_id	label_node_role_kubernetes_io	Value
m5.xlarge				amd64				rhcos				master				12
m5.xlarge				amd64				rhcos								12
```

**Comment by raptorsun** [comment]
> 1. it seems the new expr is wrong, one node is both infra/worker node
> 
> ```
> $ oc get node -l node-role.kubernetes.io/infra
> NAME                                         STATUS   ROLES          AGE    VERSION
> ip-10-0-217-149.us-east-2.compute.internal   Ready    infra,worker   129m   v1.26.2+dc93b13
> ```
> 
> search with the new expr, no infra result
> 
> ```
> label_beta_kubernetes_io_instance_type	label_kubernetes_io_arch	label_node_openshift_io_os_id	label_node_role_kubernetes_io	Value
> m5.xlarge				amd64				rhcos				master				12
> m5.xlarge				amd64				rhcos								12
> ```
> 
> 2. if node is also infra and master node
> 
> ```
> $ oc get node -l node-role.kubernetes.io/infra
> NAME                                         STATUS   ROLES                        AGE    VERSION
> ip-10-0-188-123.us-east-2.compute.internal   Ready    control-plane,infra,master   147m   v1.26.2+dc93b13
> ```
> 
> search with the new expr, no infra result either
> 
> ```
> label_beta_kubernetes_io_instance_type	label_kubernetes_io_arch	label_node_openshift_io_os_id	label_node_role_kubernetes_io	Value
> m5.xlarge				amd64				rhcos				master				12
> m5.xlarge				amd64				rhcos								12
> ```
> 
> I think we need to consider the following scenarios:
> 
> 1. some nodes are both labeled with infra and worker
> 2. some nodes are both labeled with infra and master
> 3. some nodes are only labeled with infra
>    maybe we need to talk to other developers and the bug reporter to see if the above scenarios is valid

We got confirmation on this logic to label nodes with multiple roles from reporter @joshuawilson https://issues.redhat.com/browse/OCPBUGS-10387

**Comment by raptorsun** [comment]
/hold 
still verifying labeling rules of nodes with multiple roles.

**Comment by juzhao** [comment]
bug: https://issues.redhat.com/browse/OCPBUGS-12525

**Comment by raptorsun** [comment]
> bug: https://issues.redhat.com/browse/OCPBUGS-12525

Thanks, I will see how to fix the bug.

File: test/rules/cluster_capacity_cpu.yaml [hunk 1]
```diff
@@ -64,9 +64,11 @@ tests:
       - expr: cluster:capacity_cpu_cores:sum
         eval_time: 1m
         exp_samples:
+          - labels: 'cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="e2-standard-4",label_node_role_kubernetes_io="infra",label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos"}'
+            value: 1
           - labels: 'cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="e2-standard-4",label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos"}'
-            value: 7
-  # master node capacity is listed as separate series
+            value: 6
+  # master node capacity is listed as seperate series
```
Comment (simonpasquier) [inline test/rules/cluster_capacity_cpu.yaml:71]: s/seperate/distinct/

File: test/rules/cluster_capacity_cpu.yaml [hunk 2]
```diff
@@ -37,7 +37,7 @@ tests:
         exp_samples:
           - labels: 'cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="e2-standard-4",label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos"}'
             value: 6
-  # infra node capacity is listed in the same series as workers
+  # infra node capacity is listed as different series than workers.
   # This behavior of labeling infra node without label_node_role_kubernetes_io in cluster:capacity_cpu_cores:sum is reported as a bug:
```
Comment (simonpasquier) [inline test/rules/cluster_capacity_cpu.yaml:41]: should remove this?

File: test/rules/cluster_capacity_cpu.yaml [hunk 3]
```diff
@@ -161,5 +161,36 @@ tests:
         exp_samples:
           - labels: 'cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="e2-standard-4",label_node_role_kubernetes_io="master",label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos"}'
             value: 5
+          - labels: 'cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="e2-standard-4",label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos"}'
+            value: 4
+          - labels: 'cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="e2-standard-4",label_node_role_kubernetes_io="infra",label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos"}'
+            value: 4
+  # node with both custom role ("ai-worker" in the test) is counted as worker.
```
Comment (simonpasquier) [inline test/rules/cluster_capacity_cpu.yaml:168]: `node with both custom role` did you mean `node with a custom role other than infra or master`?

Comment (raptorsun) [reply test/rules/cluster_capacity_cpu.yaml:168]: exactly, let me rephrase it in a better way.

**Comment by juzhao** [comment]
[OCPBUGS-12525](https://issues.redhat.com/browse/OCPBUGS-12525) and [OCPBUGS-10387](https://issues.redhat.com/browse/OCPBUGS-10387) are fixed
/label qe-approved

**Review by simonpasquier — COMMENTED** [review]
/lgtm
